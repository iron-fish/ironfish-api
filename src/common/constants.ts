/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { EventType } from '.prisma/client';

export const DAYS_IN_WEEK = 7;
export const MS_PER_DAY = 1000 * 60 * 60 * 24;

// Pagination limits
export const DEFAULT_LIMIT = 20;
export const MAX_LIMIT = 100;

// Event categories
export const WEEKLY_POINT_LIMITS_BY_EVENT_TYPE: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 1000,
  [EventType.BUG_CAUGHT]: 1000,
  [EventType.COMMUNITY_CONTRIBUTION]: 2000,
  [EventType.PULL_REQUEST_MERGED]: 5000,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 1000,
};

export const POINTS_PER_CATEGORY: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 100,
  [EventType.BUG_CAUGHT]: 100,
  [EventType.COMMUNITY_CONTRIBUTION]: 1000,
  [EventType.PULL_REQUEST_MERGED]: 500,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 100,
};
