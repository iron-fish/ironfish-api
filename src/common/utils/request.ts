/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Request } from 'express';

export function fetchIpAddressFromRequest(request: Request): string {
  const xForwardedFor = request.header('X-Forwarded-For');
  if (xForwardedFor) {
    const addresses = xForwardedFor.split(',');
    if (addresses.length) {
      return addresses[0].trim();
    }
  }

  // In Heroku, `X-Forwarded-For` will always be set. However, this will be used
  // as a fallback in local development.
  return request.ip;
}
