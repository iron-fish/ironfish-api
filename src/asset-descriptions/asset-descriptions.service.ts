/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable } from '@nestjs/common';
import {
  Asset,
  AssetDescription,
  AssetDescriptionType,
  Prisma,
  Transaction,
} from '@prisma/client';
import { DEFAULT_LIMIT, MAX_LIMIT } from '../common/constants';
import { SortOrder } from '../common/enums/sort-order';
import { PrismaService } from '../prisma/prisma.service';
import { BasePrismaClient } from '../prisma/types/base-prisma-client';
import { ListAssetDescriptionsOptions } from './interfaces/list-asset-descriptions-options';

@Injectable()
export class AssetDescriptionsService {
  constructor(private readonly prisma: PrismaService) {}

  async create(
    type: AssetDescriptionType,
    value: bigint,
    asset: Asset,
    transaction: Transaction,
    prisma: BasePrismaClient,
  ): Promise<AssetDescription> {
    return prisma.assetDescription.create({
      data: {
        type,
        value,
        asset_id: asset.id,
        transaction_id: transaction.id,
      },
    });
  }

  async findByTransaction(
    transaction: Transaction,
  ): Promise<AssetDescription[]> {
    return this.findByTransactionWithClient(
      transaction,
      this.prisma.readClient,
    );
  }

  async findByTransactionWithClient(
    transaction: Transaction,
    prisma: BasePrismaClient,
  ): Promise<AssetDescription[]> {
    return prisma.assetDescription.findMany({
      where: {
        transaction_id: transaction.id,
      },
    });
  }

  async deleteByTransaction(
    transaction: Transaction,
    prisma: BasePrismaClient,
  ): Promise<void> {
    await prisma.assetDescription.deleteMany({
      where: {
        transaction_id: transaction.id,
      },
    });
  }

  async list(options: ListAssetDescriptionsOptions): Promise<{
    data: AssetDescription[];
    hasNext: boolean;
    hasPrevious: boolean;
  }> {
    const cursorId = options.before ?? options.after;
    const cursor = cursorId ? { id: cursorId } : undefined;
    const direction = options.before !== undefined ? -1 : 1;
    const limit =
      direction * Math.min(MAX_LIMIT, options.limit || DEFAULT_LIMIT);
    const orderBy = { id: SortOrder.DESC };
    const skip = cursor ? 1 : 0;
    const where = {
      asset_id: options.assetId,
    };
    const data = await this.prisma.assetDescription.findMany({
      cursor,
      orderBy,
      skip,
      take: limit,
      where,
    });

    return {
      data,
      ...(await this.getListMetadata(data, where, orderBy)),
    };
  }

  private async getListMetadata(
    data: AssetDescription[],
    where: Prisma.AssetDescriptionWhereInput,
    orderBy: Prisma.Enumerable<Prisma.AssetDescriptionOrderByWithRelationInput>,
  ): Promise<{ hasNext: boolean; hasPrevious: boolean }> {
    const { length } = data;
    if (length === 0) {
      return {
        hasNext: false,
        hasPrevious: false,
      };
    }
    const nextRecords = await this.prisma.assetDescription.findMany({
      where,
      orderBy,
      cursor: { id: data[length - 1].id },
      skip: 1,
      take: 1,
    });
    const previousRecords = await this.prisma.assetDescription.findMany({
      where,
      orderBy,
      cursor: { id: data[0].id },
      skip: 1,
      take: -1,
    });
    return {
      hasNext: nextRecords.length > 0,
      hasPrevious: previousRecords.length > 0,
    };
  }
}
