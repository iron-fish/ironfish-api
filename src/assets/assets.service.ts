/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable, NotFoundException } from '@nestjs/common';
import { DEFAULT_LIMIT, MAX_LIMIT } from '../common/constants';
import { SortOrder } from '../common/enums/sort-order';
import { PrismaService } from '../prisma/prisma.service';
import { BasePrismaClient } from '../prisma/types/base-prisma-client';
import { CreateAssetOptions } from './interfaces/create-asset-options';
import { ListAssetsOptions } from './interfaces/list-assets-optionts';
import { Asset, Prisma, Transaction } from '.prisma/client';

@Injectable()
export class AssetsService {
  constructor(private readonly prisma: PrismaService) {}

  async findOrThrow(id: number): Promise<Asset> {
    return this.findOrThrowWithClient(id, this.prisma.readClient);
  }

  async findOrThrowWithClient(
    id: number,
    prisma: BasePrismaClient,
  ): Promise<Asset> {
    const record = await prisma.asset.findUnique({
      where: {
        id,
      },
    });
    if (!record) {
      throw new NotFoundException();
    }
    return record;
  }

  async findByIdentifierOrThrow(identifier: string): Promise<Asset> {
    return this.findByIdentifierOrThrowWithClient(identifier, this.prisma);
  }

  async findByIdentifierOrThrowWithClient(
    identifier: string,
    prisma: BasePrismaClient,
  ): Promise<Asset> {
    const record = await prisma.asset.findUnique({
      where: {
        identifier,
      },
    });
    if (!record) {
      throw new NotFoundException();
    }
    return record;
  }

  async upsert(
    options: CreateAssetOptions,
    transaction: Transaction,
    prisma: BasePrismaClient,
  ): Promise<Asset> {
    return prisma.asset.upsert({
      create: {
        identifier: options.identifier,
        metadata: options.metadata,
        name: options.name,
        owner: options.owner,
        supply: BigInt(0),
        created_transaction_id: transaction.id,
      },
      update: {},
      where: {
        identifier: options.identifier,
      },
    });
  }

  async updateSupply(
    asset: Asset,
    delta: bigint,
    prisma: BasePrismaClient,
  ): Promise<Asset> {
    return prisma.asset.update({
      data: {
        supply: {
          increment: delta,
        },
      },
      where: {
        id: asset.id,
      },
    });
  }

  async list(options: ListAssetsOptions): Promise<{
    data: Asset[];
    hasNext: boolean;
    hasPrevious: boolean;
  }> {
    const cursorId = options.before ?? options.after;
    const cursor = cursorId ? { id: cursorId } : undefined;
    const direction = options.before !== undefined ? -1 : 1;
    const limit =
      direction * Math.min(MAX_LIMIT, options.limit || DEFAULT_LIMIT);
    const orderBy = { id: SortOrder.DESC };
    const skip = cursor ? 1 : 0;

    let where: Prisma.AssetWhereInput = {};
    if (options.search) {
      where = {
        name: {
          contains: options.search,
        },
      };
    }

    const data = await this.prisma.readClient.asset.findMany({
      cursor,
      orderBy,
      skip,
      take: limit,
      where,
    });

    return {
      data,
      ...(await this.getListMetadata(data, where, orderBy)),
    };
  }

  private async getListMetadata(
    data: Asset[],
    where: Prisma.AssetWhereInput,
    orderBy: Prisma.Enumerable<Prisma.AssetOrderByWithRelationInput>,
  ): Promise<{ hasNext: boolean; hasPrevious: boolean }> {
    const { length } = data;
    if (length === 0) {
      return {
        hasNext: false,
        hasPrevious: false,
      };
    }
    const nextRecords = await this.prisma.readClient.asset.findMany({
      where,
      orderBy,
      cursor: { id: data[length - 1].id },
      skip: 1,
      take: 1,
    });
    const previousRecords = await this.prisma.readClient.asset.findMany({
      where,
      orderBy,
      cursor: { id: data[0].id },
      skip: 1,
      take: -1,
    });
    return {
      hasNext: nextRecords.length > 0,
      hasPrevious: previousRecords.length > 0,
    };
  }
}
