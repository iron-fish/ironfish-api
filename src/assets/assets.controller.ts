/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  Controller,
  Get,
  Header,
  HttpStatus,
  NotFoundException,
  Query,
  Req,
  Res,
  ValidationPipe,
} from '@nestjs/common';
import { ApiOperation, ApiTags } from '@nestjs/swagger';
import { Request, Response } from 'express';
import { BlocksTransactionsService } from '../blocks-transactions/blocks-transactions.service';
import { handleIfModifiedSince } from '../common/utils/request';
import { TransactionsService } from '../transactions/transactions.service';
import { AssetsService } from './assets.service';
import { AssetQueryDto } from './dto/asset-query-dto';
import { AssetsQueryDto } from './dto/assets-query.dto';
import { SerializedAsset } from './interfaces/serialized-asset';

@ApiTags('Assets')
@Controller('assets')
export class AssetsController {
  constructor(
    private readonly assetsService: AssetsService,
    private readonly blocksTransactionsService: BlocksTransactionsService,
    private readonly transactionsService: TransactionsService,
  ) {}

  @ApiOperation({ summary: 'Gets an asset by identifier' })
  @Get('find')
  async find(
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { id }: AssetQueryDto,
  ): Promise<SerializedAsset> {
    const asset = await this.assetsService.findByIdentifierOrThrow(id);
    const transaction = await this.transactionsService.findOrThrow(
      asset.created_transaction_id,
    );

    const blocks = await this.blocksTransactionsService.findBlocksByTransaction(
      transaction,
    );
    const block = blocks.find((block) => block.main);
    if (!block) {
      throw new NotFoundException();
    }

    return {
      object: 'asset',
      created_transaction_hash: transaction.hash,
      created_transaction_timestamp: block.timestamp.toISOString(),
      id: asset.id,
      identifier: asset.identifier,
      metadata: asset.metadata,
      name: asset.name,
      owner: asset.owner,
      supply: asset.supply.toString(),
      verified_at: asset.verified_at?.toISOString() ?? null,
    };
  }

  @ApiOperation({ summary: 'Lists assets' })
  @Get()
  @Header('Cache-Control', 's-maxage=60, stale-if-error=60')
  async list(
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    dto: AssetsQueryDto,
    @Req() req: Request,
    @Res() res: Response,
  ): Promise<void> {
    const lastUpdate = await this.assetsService.lastUpdate();
    if (lastUpdate) {
      handleIfModifiedSince(lastUpdate, req, res);
    }

    const { data, hasNext, hasPrevious } = await this.assetsService.list(dto);

    const serializedData: SerializedAsset[] = [];
    for (const asset of data) {
      const transaction = await this.transactionsService.findOrThrow(
        asset.created_transaction_id,
      );

      const blocks =
        await this.blocksTransactionsService.findBlocksByTransaction(
          transaction,
        );
      const block = blocks.find((block) => block.main);
      if (!block) {
        continue;
      }

      serializedData.push({
        object: 'asset',
        created_transaction_hash: transaction.hash,
        created_transaction_timestamp: block.timestamp.toISOString(),
        id: asset.id,
        identifier: asset.identifier,
        metadata: asset.metadata,
        name: asset.name,
        owner: asset.owner,
        supply: asset.supply.toString(),
        verified_at: asset.verified_at?.toISOString() ?? null,
      });
    }

    res.json({
      object: 'list',
      data: serializedData,
      metadata: {
        has_next: hasNext,
        has_previous: hasPrevious,
      },
    });
  }
}
