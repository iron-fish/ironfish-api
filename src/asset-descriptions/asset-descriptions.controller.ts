/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  Controller,
  Get,
  HttpStatus,
  Query,
  ValidationPipe,
} from '@nestjs/common';
import { ApiOperation, ApiTags } from '@nestjs/swagger';
import { AssetsService } from '../assets/assets.service';
import { BlocksTransactionsService } from '../blocks-transactions/blocks-transactions.service';
import { PaginatedList } from '../common/interfaces/paginated-list';
import { TransactionsService } from '../transactions/transactions.service';
import { AssetDescriptionsService } from './asset-descriptions.service';
import { AssetDescriptionsQueryDto } from './dto/asset-descriptions-query.dto';
import { SerializedAssetDescription } from './interfaces/serialized-asset-description';
import { serializedAssetDescriptionWithTimestampFromRecord } from './utils/asset-descriptions.translator';

@ApiTags('AssetDescriptions')
@Controller('asset_descriptions')
export class AssetDescriptionsController {
  constructor(
    private readonly assetDescriptionsService: AssetDescriptionsService,
    private readonly assetsService: AssetsService,
    private readonly blocksTransactionsService: BlocksTransactionsService,
    private readonly transactionsService: TransactionsService,
  ) {}

  @ApiOperation({ summary: 'Lists asset descriptions' })
  @Get()
  async list(
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { asset: assetIdentifier, after, before, limit }: AssetDescriptionsQueryDto,
  ): Promise<PaginatedList<SerializedAssetDescription>> {
    const asset = await this.assetsService.findByIdentifierOrThrow(
      assetIdentifier,
    );
    const { data, hasNext, hasPrevious } =
      await this.assetDescriptionsService.list({
        assetId: asset.id,
        after,
        before,
        limit,
      });

    const serializedData: SerializedAssetDescription[] = [];
    for (const record of data) {
      const transaction = await this.transactionsService.findOrThrow(
        record.transaction_id,
      );
      const blocks =
        await this.blocksTransactionsService.findBlocksByTransaction(
          transaction,
        );
      // Skip non main chain descriptions
      const block = blocks.find((block) => block.main);
      if (!block) {
        continue;
      }

      serializedData.push(
        serializedAssetDescriptionWithTimestampFromRecord(
          record,
          transaction,
          block,
        ),
      );
    }

    return {
      object: 'list',
      data: serializedData,
      metadata: {
        has_next: hasNext,
        has_previous: hasPrevious,
      },
    };
  }
}
