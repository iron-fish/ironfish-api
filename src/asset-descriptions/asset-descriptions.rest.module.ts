/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { AssetsModule } from '../assets/assets.module';
import { TransactionsModule } from '../transactions/transactions.module';
import { AssetDescriptionsController } from './asset-descriptions.controller';
import { AssetDescriptionsModule } from './asset-descriptions.module';

@Module({
  controllers: [AssetDescriptionsController],
  imports: [AssetDescriptionsModule, AssetsModule, TransactionsModule],
})
export class AssetDescriptionsRestModule {}
