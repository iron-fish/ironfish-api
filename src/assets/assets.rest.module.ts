/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { BlocksTransactionsModule } from '../blocks-transactions/blocks-transactions.module';
import { TransactionsModule } from '../transactions/transactions.module';
import { AssetsController } from './assets.controller';
import { AssetsModule } from './assets.module';

@Module({
  controllers: [AssetsController],
  imports: [AssetsModule, BlocksTransactionsModule, TransactionsModule],
})
export class AssetsRestModule {}
