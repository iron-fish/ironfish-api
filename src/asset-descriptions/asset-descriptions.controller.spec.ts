/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { HttpStatus, INestApplication } from '@nestjs/common';
import { AssetDescriptionType } from '@prisma/client';
import request from 'supertest';
import { v4 as uuid } from 'uuid';
import { AssetsService } from '../assets/assets.service';
import { PrismaService } from '../prisma/prisma.service';
import { bootstrapTestApp } from '../test/test-app';
import { TransactionsService } from '../transactions/transactions.service';
import { AssetDescriptionsService } from './asset-descriptions.service';

describe('AssetDescriptionsController', () => {
  let app: INestApplication;
  let assetDescriptionsService: AssetDescriptionsService;
  let assetsService: AssetsService;
  let prisma: PrismaService;
  let transactionsService: TransactionsService;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    assetDescriptionsService = app.get(AssetDescriptionsService);
    assetsService = app.get(AssetsService);
    prisma = app.get(PrismaService);
    transactionsService = app.get(TransactionsService);
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('GET /asset_descriptions', () => {
    describe('with an invalid query', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/asset_descriptions')
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a missing record', () => {
      it('returns a 404', async () => {
        await request(app.getHttpServer())
          .get('/asset_descriptions')
          .query({ asset: 'asdf' })
          .expect(HttpStatus.NOT_FOUND);
      });
    });

    describe('with a valid identifier', () => {
      it('returns the asset descriptions', async () => {
        const transaction = (
          await transactionsService.createMany([
            {
              fee: 0,
              hash: uuid(),
              notes: [],
              size: 0,
              spends: [],
            },
          ])
        )[0];

        const asset = await assetsService.upsert(
          {
            identifier: uuid(),
            metadata: uuid(),
            name: uuid(),
            owner: uuid(),
          },
          transaction,
          prisma,
        );

        const assetDescription = await assetDescriptionsService.create(
          AssetDescriptionType.MINT,
          BigInt(1),
          asset,
          transaction,
          prisma,
        );

        const secondTransaction = (
          await transactionsService.createMany([
            {
              fee: 0,
              hash: 'foo',
              notes: [],
              size: 0,
              spends: [],
            },
          ])
        )[0];
        const secondAssetDescription = await assetDescriptionsService.create(
          AssetDescriptionType.BURN,
          BigInt(1),
          asset,
          secondTransaction,
          prisma,
        );

        const { body } = await request(app.getHttpServer())
          .get('/asset_descriptions')
          .query({ asset: asset.identifier })
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          object: 'list',
          data: [
            {
              object: 'asset_description',
              id: secondAssetDescription.id,
              transaction_hash: secondTransaction.hash,
              type: AssetDescriptionType.BURN,
              value: secondAssetDescription.value.toString(),
            },
            {
              object: 'asset_description',
              id: assetDescription.id,
              transaction_hash: transaction.hash,
              type: AssetDescriptionType.MINT,
              value: assetDescription.value.toString(),
            },
          ],
          metadata: {
            has_next: false,
            has_previous: false,
          },
        });
      });
    });
  });
});
