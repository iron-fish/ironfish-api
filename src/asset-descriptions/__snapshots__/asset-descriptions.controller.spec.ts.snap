// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetDescriptionsController GET /asset_descriptions with an invalid query returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "\\"asset\\" required to query for asset descriptions",
    "asset must be a string",
  ],
  "statusCode": 422,
}
`;
